### ISSUE TITLE
ContextualVersionConflict with idna

### ISSUE THREAD

**Issue by rfleschenberg** [certbot/certbot#5015]
## My operating system is (include version):

Can reproduce on Debian 8 and Ubuntu 16.04, did not test other systems, but it seems likely all OS are affected.

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):

pip

## I ran this command and it produced this output:

`certbot`:

```
An unexpected error occurred:
ContextualVersionConflict: (idna 2.6 (/tmp/letest/lib/python2.7/site-packages), Requirement.parse('idna<2.6,>=2.5'), set(['requests']))
Please see the logfile '/tmp/tmpFj1SPt' for more details.
```

## Certbot's behavior differed from what I expected because:

It should work ;)

## Here is the full, reproducible output:

```
rene@rene-desktop /tmp % virtualenv letest
Running virtualenv with interpreter /usr/bin/python2
New python executable in /tmp/letest/bin/python2
Also creating executable in /tmp/letest/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.
rene@rene-desktop /tmp % source letest/bin/activate
(letest) rene@rene-desktop /tmp % pip install certbot
Collecting certbot
  Downloading certbot-0.17.0-py2.py3-none-any.whl (269kB)
    100% |████████████████████████████████| 276kB 807kB/s 
Collecting zope.interface (from certbot)
  Downloading zope.interface-4.4.2-cp27-cp27mu-manylinux1_x86_64.whl (170kB)
    100% |████████████████████████████████| 174kB 1.3MB/s 
Collecting pyrfc3339 (from certbot)
  Downloading pyRFC3339-1.0-py2.py3-none-any.whl
Requirement already satisfied: setuptools>=1.0 in ./letest/lib/python2.7/site-packages (from certbot)
Collecting zope.component (from certbot)
  Downloading zope.component-4.4.0-py2.py3-none-any.whl (70kB)
    100% |████████████████████████████████| 71kB 1.6MB/s 
Collecting pytz (from certbot)
  Using cached pytz-2017.2-py2.py3-none-any.whl
Collecting ConfigArgParse>=0.9.3 (from certbot)
  Downloading ConfigArgParse-0.12.0.tar.gz (41kB)
    100% |████████████████████████████████| 51kB 2.4MB/s 
Collecting configobj (from certbot)
  Downloading configobj-5.0.6.tar.gz
Collecting six (from certbot)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting cryptography>=1.2 (from certbot)
  Downloading cryptography-2.0.3-cp27-cp27mu-manylinux1_x86_64.whl (2.2MB)
    100% |████████████████████████████████| 2.2MB 435kB/s 
Collecting PyOpenSSL (from certbot)
  Downloading pyOpenSSL-17.2.0-py2.py3-none-any.whl (52kB)
    100% |████████████████████████████████| 61kB 2.0MB/s 
Collecting mock (from certbot)
  Using cached mock-2.0.0-py2.py3-none-any.whl
Collecting parsedatetime>=1.3 (from certbot)
  Downloading parsedatetime-2.4-py2-none-any.whl (40kB)
    100% |████████████████████████████████| 40kB 1.3MB/s 
Collecting acme==0.17.0 (from certbot)
  Downloading acme-0.17.0-py2.py3-none-any.whl (99kB)
    100% |████████████████████████████████| 102kB 1.5MB/s 
Collecting zope.event (from zope.component->certbot)
  Downloading zope.event-4.3.0-py2.py3-none-any.whl
Collecting ipaddress (from cryptography>=1.2->certbot)
  Using cached ipaddress-1.0.18-py2-none-any.whl
Collecting idna>=2.1 (from cryptography>=1.2->certbot)
  Downloading idna-2.6-py2.py3-none-any.whl (56kB)
    100% |████████████████████████████████| 61kB 1.7MB/s 
Collecting asn1crypto>=0.21.0 (from cryptography>=1.2->certbot)
  Using cached asn1crypto-0.22.0-py2.py3-none-any.whl
Collecting enum34 (from cryptography>=1.2->certbot)
  Using cached enum34-1.1.6-py2-none-any.whl
Collecting cffi>=1.7 (from cryptography>=1.2->certbot)
  Using cached cffi-1.10.0-cp27-cp27mu-manylinux1_x86_64.whl
Collecting funcsigs>=1; python_version < "3.3" (from mock->certbot)
  Using cached funcsigs-1.0.2-py2.py3-none-any.whl
Collecting pbr>=0.11 (from mock->certbot)
  Using cached pbr-3.1.1-py2.py3-none-any.whl
Collecting future (from parsedatetime>=1.3->certbot)
  Downloading future-0.16.0.tar.gz (824kB)
    100% |████████████████████████████████| 829kB 792kB/s 
Collecting requests[security]>=2.4.1 (from acme==0.17.0->certbot)
  Using cached requests-2.18.3-py2.py3-none-any.whl
Collecting pycparser (from cffi>=1.7->cryptography>=1.2->certbot)
Collecting chardet<3.1.0,>=3.0.2 (from requests[security]>=2.4.1->acme==0.17.0->certbot)
  Using cached chardet-3.0.4-py2.py3-none-any.whl
Collecting certifi>=2017.4.17 (from requests[security]>=2.4.1->acme==0.17.0->certbot)
  Using cached certifi-2017.7.27.1-py2.py3-none-any.whl
Collecting urllib3<1.23,>=1.21.1 (from requests[security]>=2.4.1->acme==0.17.0->certbot)
  Using cached urllib3-1.22-py2.py3-none-any.whl
Building wheels for collected packages: ConfigArgParse, configobj, future
  Running setup.py bdist_wheel for ConfigArgParse ... done
  Stored in directory: /home/rene/.cache/pip/wheels/6a/40/6b/6d9924b7757fff15a7296256f9d6d27a9600beabd387f057af
  Running setup.py bdist_wheel for configobj ... done
  Stored in directory: /home/rene/.cache/pip/wheels/87/76/48/1564f8466fbd36402af5ac4972ffb56a6ef7f143892ef57fe5
  Running setup.py bdist_wheel for future ... done
  Stored in directory: /home/rene/.cache/pip/wheels/c2/50/7c/0d83b4baac4f63ff7a765bd16390d2ab43c93587fac9d6017a
Successfully built ConfigArgParse configobj future
Installing collected packages: zope.interface, pytz, pyrfc3339, zope.event, zope.component, ConfigArgParse, six, configobj, ipaddress, idna, asn1crypto, enum34, pycparser, cffi, cryptography, PyOpenSSL, funcsigs, pbr, mock, future, parsedatetime, chardet, certifi, urllib3, requests, acme, certbot
Successfully installed ConfigArgParse-0.12.0 PyOpenSSL-17.2.0 acme-0.17.0 asn1crypto-0.22.0 certbot-0.17.0 certifi-2017.7.27.1 cffi-1.10.0 chardet-3.0.4 configobj-5.0.6 cryptography-2.0.3 enum34-1.1.6 funcsigs-1.0.2 future-0.16.0 idna-2.6 ipaddress-1.0.18 mock-2.0.0 parsedatetime-2.4 pbr-3.1.1 pycparser-2.18 pyrfc3339-1.0 pytz-2017.2 requests-2.18.3 six-1.10.0 urllib3-1.22 zope.component-4.4.0 zope.event-4.3.0 zope.interface-4.4.2
(letest) rene@rene-desktop /tmp % certbot
An unexpected error occurred:
ContextualVersionConflict: (idna 2.6 (/tmp/letest/lib/python2.7/site-packages), Requirement.parse('idna<2.6,>=2.5'), set(['requests']))
Please see the logfile '/tmp/tmpFj1SPt' for more details.
(letest) rene@rene-desktop /tmp %
```

Manually downgrading `idna` to `2.5` seems to work around the problem (thanks to Peng on #letsencrypt for the hint).
```
(letest) rene@rene-desktop /tmp % pip install idna\<2.6
Collecting idna<2.6
  Using cached idna-2.5-py2.py3-none-any.whl
Installing collected packages: idna
  Found existing installation: idna 2.6
    Uninstalling idna-2.6:
      Successfully uninstalled idna-2.6
Successfully installed idna-2.5
(letest) rene@rene-desktop /tmp % certbot
The following error was encountered:
[Errno 13] Permission denied: '/var/log/letsencrypt'
Either run as root, or set --config-dir, --work-dir, and --logs-dir to writeable paths.
```

**Comment by SwartzCr** [comment]
@rfleschenberg so we don't actually support install via pip - on Debian 8 you should be installing Certbot via apt with jessie-backports, as explained here: https://certbot.eff.org/#debianjessie-apache
I'm curious if other people who have run into this issue also installed via pip

**Comment by melato** [comment]
I get this error on Alpine edge, installed as an LXD container on Ubuntu.  I have gotten missing python dependencies on Alpine edge before, which were promptly fixed by the Alpine team.  But this is somewhat different, so I'm posting instructions here:
On Ubuntu 16.04, create an Alpine edge container called "edge" and start an ash shell on it:
```
lxc launch images:alpine/edge edge
lxc exec edge ash

```
On the newly created Alpine container:
```
# apk update
# apk upgrade
# apk add certbot
# certbot --version

```
Result:
pkg_resources.ContextualVersionConflict: (idna 2.6 (/usr/lib/python2.7/site-packages), Requirement.parse('idna<2.6,>=2.5'), set(['requests']))

I verified that it works fine on Alpine 3.6, which has certbot version 0.14.0, but I use Alpine edge to get a more recent version of certbot (It was 0.17.0 before the broken upgrade).

**Comment by bmw** [comment]
It looks like Alpine edge currently doesn't contain a version of `idna` that satisfies `requests`.
```
$ sudo docker run --rm -it alpine:edge sh -c 'apk update >/dev/null 2>&1; apk add py-requests'
(1/5) Installing py-chardet (3.0.4-r0)
(2/5) Installing py-idna (2.6-r0)
(3/5) Installing py-certifi (2017.4.17-r0)
(4/5) Installing py-urllib3 (1.22-r0)
(5/5) Installing py-requests (2.18.3-r0)
OK: 4 MiB in 16 packages
```
And if you look at Python request 2.18.3, they require [idna <2.6](https://github.com/requests/requests/blob/v2.18.3/setup.py#L46).

**Comment by bcutter** [comment]
Same here. 
```
An unexpected error occurred:
ContextualVersionConflict: (idna 2.6 (/root/.local/share/letsencrypt/lib/python2.7/site-packages), Requirement.parse('idna<2.6,>=2.5'), set(['requests']))
```

And "/letsencrypt/certbot.log" gives
```
Traceback (most recent call last):
  File "/root/.local/share/letsencrypt/bin/letsencrypt", line 11, in <module>
    sys.exit(main())
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/certbot/main.py", line 709, in main
    cli.possible_deprecation_warning(config)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/certbot/cli.py", line 160, in possible_deprecation_warning
    logger.warn("You are running with an old copy of letsencrypt-auto that does "
  File "/usr/lib/python2.7/logging/__init__.py", line 1172, in warning
    self._log(WARNING, msg, args, **kwargs)
  File "/usr/lib/python2.7/logging/__init__.py", line 1279, in _log
    self.handle(record)
  File "/usr/lib/python2.7/logging/__init__.py", line 1289, in handle
    self.callHandlers(record)
  File "/usr/lib/python2.7/logging/__init__.py", line 1329, in callHandlers
    hdlr.handle(record)
  File "/usr/lib/python2.7/logging/__init__.py", line 757, in handle
    self.emit(record)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/certbot/log.py", line 64, in emit
    self.width + self.PADDING_WIDTH)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/dialog.py", line 2748, in infobox
    kwargs)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/dialog.py", line 1767, in _widget_with_no_output
    widget_name, output))
PythonDialogBug
```

Is it because Python version is too new? Not sure if it was updated since the last cron based certificate renewal run.

**Comment by bmw** [comment]
@bcutter, I think you're using a very old version of `letsencrypt-auto`. I recommend you download a new version using the instructions [here](https://certbot.eff.org/docs/install.html#certbot-auto).

**Comment by pombredanne** [comment]
@SwartzCr you wrote:

> I'm curious if other people who have run into this issue also installed via pip

I have just hit the issue yesterday. Same thing as others: I have certbot installed via pip (through ansible though it does not matter) on Ubuntu 16.x servers. I resolved this with an uninstall of idna and a manual install of idna==2.5


> so we don't actually support install via pip

Then cerbot should not be pushed to Pypi maybe?

Each time I updated cerbot over the last year or so I ended up with some dependency issues.
Being somewhat Python-literate, I would think that vendoring or bundling all your dependencies is the only sane way to avoid these class of problems for such a tool.

**Comment by pombredanne** [comment]
FWIW a simple fresh pip install in a venv fails nicely:
```
$ pip install certbot
[...]
$ certbot --help
An unexpected error occurred:
ContextualVersionConflict: (idna 2.6 (/tmp/cerbottst/tmp/lib/python2.7/site-packages), Requirement.parse('idna<2.6,>=2.5'), set(['requests']))
Please see the logfile '/tmp/tmpMpUHeO' for more details.
```

So another way to ensure that you have proper deps on a plain pip install is tu setup stricter version ranges in you setup.py.... but that would eventually make it harder to use cerbot as a library if this is a use case at all.

Yet another way is to have two packages: the regular one as today and a second one that provides no code per se but only stricly pinned deps from https://github.com/certbot/certbot/blob/16ed1413013572880f79d2460517d8e85f202e27/certbot-auto#L719 and the regular package as install required... That second one would be a kinda wrapper that could be generated and would suitable for pip installation when using cerbot as CLI.... The main one (when installed without certbot-auto) could be used when used a library.

**Comment by HackintoshHD** [comment]
Just for the record, I got the same error message as @rfleschenberg above on a Debian 8.9 Jessie webserver.

I got **around it** by basically deleting `/root/.local/share/letsencrypt` (or more specifically backing it up by moving it to a new name). To avoid loosing the configuration, press 'c' to cancel the setup at the corresponding moment:

```
root@webserver:/home/administrator/letsencrypt $ ./certbot-auto renew --rsa-key-size 4096 --standalone
Error: couldn't get currently installed version for /root/.local/share/letsencrypt/bin/letsencrypt: 
An unexpected error occurred:
ContextualVersionConflict: (idna 2.6 (/root/.local/share/letsencrypt/lib/python2.7/site-packages), Requirement.parse('idna<2.6,>=2.5'), set(['requests']))
Please see the logfile '/tmp/tmpxzpbPr' for more details.

root@webserver:/home/administrator/letsencrypt $ cd /root/.local/share

root@webserver:~/.local/share $ mv letsencrypt letsencrypt-backup

root@webserver:~/.local/share $ cd /home/administrator/letsencrypt

root@webserver:/home/administrator/letsencrypt $ ./certbot-auto
Bootstrapping dependencies for Debian-based OSes... (you can skip this with --no-bootstrap)
[…]
Creating virtual environment...
Installing Python packages...
Installation succeeded.
Saving debug log to /var/log/letsencrypt/letsencrypt.log

Which names would you like to activate HTTPS for?
-------------------------------------------------------------------------------
1: your.domain1.tld
2: your.domain2.tld
-------------------------------------------------------------------------------
Select the appropriate numbers separated by commas and/or spaces, or leave input
blank to select all options shown (Enter 'c' to cancel): c
Please specify --domains, or --installer that will help in domain names autodiscovery, or --cert-name for an existing certificate name.

root@webserver:/home/administrator/letsencrypt $ ./certbot-auto renew --rsa-key-size 4096 --standalone
Saving debug log to /var/log/letsencrypt/letsencrypt.log

-------------------------------------------------------------------------------
Processing /etc/letsencrypt/renewal/your.domain1.tld.conf
-------------------------------------------------------------------------------
Cert is due for renewal, auto-renewing...
Renewing an existing certificate
Performing the following challenges:
tls-sni-01 challenge for your.domain1.tld
Waiting for verification...
Cleaning up challenges

-------------------------------------------------------------------------------
new certificate deployed without reload, fullchain is
/etc/letsencrypt/live/your.domain1.tld/fullchain.pem
-------------------------------------------------------------------------------

-------------------------------------------------------------------------------
Processing /etc/letsencrypt/renewal/your.domain2.tld.conf
-------------------------------------------------------------------------------
Cert is due for renewal, auto-renewing...
Renewing an existing certificate
Performing the following challenges:
tls-sni-01 challenge for your.domain2.tld
Waiting for verification...
Cleaning up challenges

-------------------------------------------------------------------------------
new certificate deployed without reload, fullchain is
/etc/letsencrypt/live/your.domain2.tld/fullchain.pem
-------------------------------------------------------------------------------

Congratulations, all renewals succeeded. The following certs have been renewed:
  /etc/letsencrypt/live/your.domain1.tld/fullchain.pem (success)
  /etc/letsencrypt/live/your.domain2.tld/fullchain.pem (success)
  […]
root@webserver:/home/administrator/letsencrypt $
```

**Comment by TheFarodin** [comment]
I've just run into the same error message on debian 8, however I haven't installed letsencrypt via pip.
Almost 2 years ago I installed it using the git method (since that was recommended at the time) and have been running "renew" via cron ever since.
Installing idna-2.5 didn't work for me, it would immediately be upgraded again when running letsencrypt-auto (this is why I dislike self-updating programs that don't at least give you a choice).
The only way I could get anywhere was by installing certbot from jessie-backport. It completely fails at renewing certificates and complains about having to parse version 0.16.0 config files with it's version 0.10.2, however manually issueing a new certificate still worked.

**Comment by barshinoff** [comment]
I agree a program that does self updating is ridiculous, especially when these certs expire in 3 months and you can't renew until the last 2 months. By the time you realize the update which was mandatory as part of the renewal process is not going to work you don't have much time to find a solution before the cert expires!

**Comment by jamescorkran** [comment]
I am just now running into this problem on Ubuntu 14.04 using Chef and letsencrypt-auto, I am at a loss of what to do and my certs will expire soon.
Latest Output (sanitized)

> Expected process to exit with [0], but received '1'
    ---- Begin output of if [ ! -f /var/www/letsencrypt/live/dev2.WEBSITE.com/cert.pem ] || test `find /var/www/letsencrypt/live/dev2.WEBSITE.com/cert.pem -mtime +50`; then /var/www/letsencrypt/letsencrypt-auto --agree-tos --email certmaster@WEBSITE.com --config-dir ~/letsencrypt --work-dir ~/letsencrypt --logs-dir ~/letsencrypt certonly --text --renew-by-default --webroot -w /var/www/dev2.WEBSITE.com -d dev2.WEBSITE.com; fi ----
    STDOUT: Updating letsencrypt and virtual environment dependencies.......
    Running with virtualenv: sudo /var/www/.local/share/letsencrypt/bin/letsencrypt --agree-tos --email certmaster@WEBSITE.com --config-dir /var/www/letsencrypt --work-dir /var/www/letsencrypt --logs-dir /var/www/letsencrypt certonly --text --renew-by-default --webroot -w /var/www/dev2.WEBSITE.com -d dev2.WEBSITE.com
    STDERR: An unexpected error occurred:
    ContextualVersionConflict: (idna 2.6 (/var/www/.local/share/letsencrypt/lib/python2.7/site-packages), Requirement.parse('idna<2.6,>=2.5'), set(['requests']))
    Please see the logfile '/tmp/tmprvrfdQ' for more details.
    ---- End output of if [ ! -f /var/www/letsencrypt/live/dev2.WEBSITE.com/cert.pem ] || test `find /var/www/letsencrypt/live/dev2.WEBSITE.com/cert.pem -mtime +50`; then /var/www/letsencrypt/letsencrypt-auto --agree-tos --email certmaster@WEBSITE.com --config-dir ~/letsencrypt --work-dir ~/letsencrypt --logs-dir ~/letsencrypt certonly --text --renew-by-default --webroot -w /var/www/dev2.WEBSITE.com -d dev2.WEBSITE.com; fi ----
    Ran if [ ! -f /var/www/letsencrypt/live/dev2.WEBSITE.com/cert.pem ] || test `find /var/www/letsencrypt/live/dev2.WEBSITE.com/cert.pem -mtime +50`; then /var/www/letsencrypt/letsencrypt-auto --agree-tos --email certmaster@WEBSITE.com --config-dir ~/letsencrypt --work-dir ~/letsencrypt --logs-dir ~/letsencrypt certonly --text --renew-by-default --webroot -w /var/www/dev2.WEBSITE.com -d dev2.WEBSITE.com; fi returned 1

**Comment by SwartzCr** [comment]
@barshinoff upgrading is not mandatory, you can have certbot-auto keep its version stable by providing the `--no-self-upgrade` flag.
@pombredanne we upload a version of certbot to pypi so that certbot-auto can download it. The reason we use certbot-auto to manage dependencies rather than pypi is because certbot-auto provides a method for getting required system level dependencies as well as the fact that pip doesn't actually have a scheme for dependency resolution.
@TheFarodin if you are having trouble renewing certificates can you please open a new issue for them? I'd be happy to help debug. That said the conf file version warning can be safely ignored, and you should be able to get a packaged version of 0.17.0 soon now that the debian stable freeze has ended.
@bmw or @erikrose do you think it's worth putting a requires for idna 2.5 in pip to help people using ansible?

**Comment by rdonkin** [comment]
I just had the same error using a Docker container for certbot built using the official `python:2.7` image, which is derived from `debian:jessie`.  It would be useful if there's a way of fixing this while using pip to install dependencies.  

I found certbot-auto was not very good for Docker because it wants to download things at run time rather than build time, and it's generally harder to debug when auto updates are happening - so I'd really prefer a pip-based setup.

**Comment by pombredanne** [comment]
@SwartzCr FWIW as an example of a fully pinned metapackage for certbot I just pushed this https://pypi.python.org/pypi/certbot-pinned and this https://github.com/pombredanne/certbot-pinned
e.g. a plain setup.py with all pinned deps that I can then use to pip install all the proper versions.

**Comment by erikrose** [comment]
@pombredanne The downside of the metapackage (which is very clever, btw) is you're trusting a whole mess of unverified downloads from PyPI to do the right thing as root on your box. certbot-auto contains a requirements file with sha-256 hashes of all the PyPI downloads and checks them as it downloads. If you're going to install using pip, I recommend extracting [those files](https://github.com/certbot/certbot/tree/master/letsencrypt-auto-source/pieces) and pointing pip to them: `pip install -r ...`. Obviously, it's a trust-on-first-use arrangement, but it's better than having faith in the certificate authority system, which has a zillion entities with do-anything access.

**Comment by SwartzCr** [comment]
@rdonkin is there a reason you don't want to use the official dockerhub image that installs the needed dependencies as part of the docker compose?

**Comment by pombredanne** [comment]
@erikrose Thanks!. I add not found "[those files](https://github.com/certbot/certbot/tree/master/letsencrypt-auto-source/pieces)" in plain text. That said, yes, the metapackage approach is faulty and I dived in the code of setuptools but there is no checksum verification that's possible so the approach is fraught with perils. 

FWIW, in order to solve a similar proper (e.g. ensure that a Python app of mine always installs with exact versions of dependencies that I vetted all the way) I also used another approach: 

I bundle all the dependencies as wheels or sdists in the repo itself down to pip, setuptools and virtualenv and use a configure script to install this entirely locally, getting everything from a release archive or checkout. 

This is of course a **terribly bad heresy** if this were a library but IMHO **it makes sense for an application** where you want fine and exact control of which deps are installed. IMHO certbot is NOT a library but an application so using a similar approach could allow such a fine control.

You can see the mechanics of this bundling here:
1. the configure [shell script](https://github.com/nexB/scancode-toolkit/blob/develop/configure) at the root that handles the bootstrapping (this is for POSIX: there is also a [configure.bat](https://github.com/nexB/scancode-toolkit/blob/develop/configure.bat) for Winwoes)...
2. ... which then  delegates to this [configure.py script](https://github.com/nexB/scancode-toolkit/blob/develop/etc/configure.py) . This script starts by creating a [virtualenv](https://github.com/nexB/scancode-toolkit/blob/develop/etc/configure.py#L172)  and then [installs ](https://github.com/nexB/scancode-toolkit/blob/develop/etc/configure.py#L232)requirement [files](https://github.com/nexB/scancode-toolkit/blob/develop/etc/conf/base.txt) without calling Pypi, without using a cache or anything external, and optionally run [shell scripts or python scripts](https://github.com/nexB/scancode-toolkit/blob/develop/etc/configure.py#L243) organized in a [directory tree](https://github.com/nexB/scancode-toolkit/tree/develop/etc/conf) to have various config possible (e.g. devel vs prod, etc) 
3. ... and all the deps are obtained locally from a [thirdparty dir](https://github.com/nexB/scancode-toolkit/tree/develop/thirdparty/prod)

At a high level this is very loosely equivalent in a much simplified way to what certbot-auto does with one major difference: all deps archives are BUNDLED. So I only need to trust (and verify) the integrity of a single release download archive. And since it runs pip, its requirements files could contain SHA256 for verification. And there are fewer moving parts since there is only a single initial download needed.

I am not saying that cerbot should use this script! but I am advocating for the bundling of deps to simplify dealing with application dependencies.

**Comment by rdonkin** [comment]
@SwartzCr Maybe I should have started with the certbot [Docker image](https://github.com/certbot/certbot/blob/master/Dockerfile).  However, the fact  thatit uses Alpine is a bit of a problem given some local DNS config issues - the `libmusl` libc in Alpine [does DNS lookups in parallel](https://github.com/gliderlabs/docker-alpine/blob/master/docs/caveats.md#dns) across all DNS servers, which causes problems in our local setup.  So I would have had to change it anyway.

Given that certbot is in PyPI, I think it's reasonable to expect a simple `pip` based install to work - in fact I tried certbot-auto and using debian-jessie backports, which both failed to work in different ways, before going for the pip install.

**Comment by erikrose** [comment]
@pombredanne We did consider a lot of alternatives, including bundling. certbot, like your tool, has to run on a variety of platforms and architectures. Since not all of our dependencies are pure Python, we'd either have to release a multiplicity of bundles (which is the tack I see you took) or else one very large one which contains all possible builds of, say, [cryptography](https://pypi.python.org/pypi/cryptography/), each of which is 1-2MB. I see you went ahead and built bundles for each platform and count on users to know which to download. This actually seems like more work and more opportunity for error than just delegating that problem to pip and PyPI. (It also doesn't hurt that PyPI has its own CDN and so handles the distribution of the great majority of our bits for us.) In any case, we'd still need a self-updating certbot-auto-like script to install the requisite OS packages—we'd never be rid of that layer. But I'm all ears for an approach that truly simplifies things! Maybe I'm missing something.

Quasi disclaimer: I'm the author of pip's hash-checking features, but I intended them for precisely such cases. Cheers!

**Comment by rfleschenberg** [comment]
> @bmw or @erikrose do you think it's worth putting a requires for idna 2.5 in pip to help people using ansible?

My POV as an Ansible user: It's not a problem to install distribution packages with Ansible. I will update the affected Ansible roles to install the .deb package. I just was not aware of the fact that a .deb package exists and that installing via pip is not supported. Maybe a warning can be added in an appropriate place?

**Comment by SwartzCr** [comment]
@rfleschenberg out of curiosity where did you get the ansible recipe(?) that suggested using pip? It seems like you're not the only one who chose to do it that way. Is that just standard practice for ansible, or is there a tutorial somewhere that recommends that?

**Comment by consense** [comment]
@SwartzCr had the same problem as rfleschenberg using https://galaxy.ansible.com/thefinn93/letsencrypt/

**Comment by bcutter** [comment]
**A short user feedback/update:**
So, no idea WHY, but when I today ran my core command for certificate renewal manually, everything was working fine. After that I did a successful second renew by using my standard script in a cron.

I also skipped my initial idea of updating my letsencrypt-auto to certbot-auto --> never change a running system, right.

**Comment by nbering** [comment]
I've also noticed that it seems to have resolved itself. Not sure if there was an intentional fix but the issue wasn't closed, or if someone just updated something with no idea that this issue was even here.

**Comment by rdonkin** [comment]
> @rfleschenberg out of curiosity where did you get the ansible recipe(?) that suggested using pip? It seems like you're not the only one who chose to do it that way. Is that just standard practice for ansible, or is there a tutorial somewhere that recommends that?

Ansible can be used to install anything including OS packages, Docker images or Python packages.  Perhaps it's just that Python users expect that Python packages published on PyPI can be installed using `pip` - if this isn't a recommended install approach, that should be clearly documented.

In my case I had problems with Jessie backports, and certbot-auto had some limitations, so pip seemed like the best solution, and it has worked fine so far.

**Comment by SwartzCr** [comment]
sounds like this is resolved (likely upstream) closing

**Comment by rfleschenberg** [comment]
@SwartzCr Just FYI, I think there are two main reasons why people use pip:

1) For a Python developer who knows that certbot is written in Python, `pip install certbot` is a somewhat natural thing to do. Distro packages are often very outdated, so many developers are basically conditioned to get Python tools from PyPi by default.

2) In the past, certbot was not trivially available on Debian Jessie (at the very least you have to enable backports, not sure for how long the backport has been available).

**Comment by praetp** [comment]
Saw the issue again on Raspbian stretch...

**Comment by philrice** [comment]
I just encountered this issue after installing certbot inside an nginx official container running on Alpine. It installed the dependencies which included python 2.7 but then has a conflict as it needs 'idna <2.7,>=2.5'

**Comment by vir2l** [comment]
I hope certbot client will get an update on this, meanwhile I installed idna-2.6 to get it running.
On alpine machines (and I am sure on others) you need to ensure that "py-pip" is installed. After that you simply need to `pip install 'idna<2.7' --force-reinstall` this downgrades idna.

**Comment by Emerson** [comment]
Same issue here on Ubuntu. Ended up just using acme.sh, was much easier.

**Comment by hussfelt** [comment]
This is still an issue.
https://community.letsencrypt.org/t/renew-certificates-contextualversionconflict/39826/17
Had to force-install another package like @vir2l.

**Comment by groenator** [comment]
Today I got the update and now I cannot issue a new certificate. The error is complaining about idna>2.8 and requests. Request 2.20.1 is not compatible with idna=2.8, but even if the version of idna is downgraded to 2.7, the issue is still not resolved. 

`sudo pip install 'idna>=2.1'
Collecting idna>=2.1
  Using cached https://files.pythonhosted.org/packages/14/2c/cd551d81dbe15200be1cf41cd03869a46fe7226e7450af7a6545bfc474c9/idna-2.8-py2.py3-none-any.whl`

requests 2.20.1 has requirement idna<2.8,>=2.5, but you'll have idna 2.8 which is incompatible.

When I upgrade request version, it downgraded idna to 2.7. The request version  stayed the same 

```
[centos@ ~]$ sudo /opt/certbot/bin/certbot certonly
An unexpected error occurred:
ContextualVersionConflict: (idna 2.8 (/opt/certbot/lib/python2.7/site-packages), Requirement.parse('idna<2.8,>=2.5'), set(['requests']))
Please see the logfile '/tmp/tmpPng4OC' for more details.
```

```
[centos@ ~]$ sudo /opt/certbot/bin/certbot certonly --help
An unexpected error occurred:
ContextualVersionConflict: (idna 2.8 (/opt/certbot/lib/python2.7/site-packages), Requirement.parse('idna<2.8,>=2.5'), set(['requests']))
```


List of pip packages installed:

```
[centos@ ~]$ sudo pip list
Package                          Version
-------------------------------- ----------
acme                             0.29.1
ansible                          2.7.4
ansible-modules-hashivault       3.10.1
asn1crypto                       0.24.0
awscli                           1.16.70
Babel                            0.9.6
backports.ssl-match-hostname     3.5.0.1
bcrypt                           3.1.4
botocore                         1.12.60
certbot                          0.27.1
certifi                          2018.11.29
cffi                             1.11.5
chardet                          3.0.4
cloud-init                       18.2
colorama                         0.3.9
ConfigArgParse                   0.13.0
configobj                        4.7.2
cryptography                     2.4.2
decorator                        3.4.0
docutils                         0.14
enum34                           1.1.6
funcsigs                         1.0.2
future                           0.17.1
futures                          3.2.0
httplib2                         0.9.2
hvac                             0.7.0
idna                             2.7
iniparse                         0.4
ipaddress                        1.0.22
IPy                              0.75
Jinja2                           2.10
jmespath                         0.9.3
josepy                           1.1.0
jsonpatch                        1.2
jsonpointer                      1.9
kitchen                          1.1.1
MarkupSafe                       1.1.0
mock                             2.0.0
ndg-httpsclient                  0.3.2
paramiko                         2.4.2
parsedatetime                    2.4
passlib                          1.7.1
pbr                              5.1.1
perf                             0.1
pip                              18.1
ply                              3.4
policycoreutils-default-encoding 0.1
prettytable                      0.7.2
pyasn1                           0.4.4
pycparser                        2.19
pycrypto                         2.6.1
pycurl                           7.19.0
pygobject                        3.22.0
pygpgme                          0.3
pyliblzma                        0.5.3
PyNaCl                           1.3.0
pyOpenSSL                        0.13.1
pyRFC3339                        1.1
pyserial                         2.6
python-dateutil                  2.7.5
python-keyczar                   0.71rc0
python-linux-procfs              0.4.9
pytz                             2018.7
pyudev                           0.15
pyxattr                          0.5.1
PyYAML                           3.13
requests                         2.20.1
requests-toolbelt                0.8.0
rsa                              3.4.2
s3transfer                       0.1.13
schedutils                       0.4
seobject                         0.1
sepolicy                         1.1
setuptools                       40.6.2
six                              1.11.0
slip                             0.4.0
slip.dbus                        0.4.0
urlgrabber                       3.10
urllib3                          1.24.1
virtualenv                       16.1.0
yum-metadata-parser              1.1.4
zope.component                   4.5
zope.deferredimport              4.3
zope.deprecation                 4.4.0
zope.event                       4.4
zope.hookable                    4.2.0
zope.interface                   4.6.0
zope.proxy                       4.3.1
```

**Comment by hussfelt** [comment]
@groenator if there is any help in this I had a problem where when I packaged my things I never cleared out the old packages. That meant that both idna 2.8 and idna 2.7 (which I replaced it with) was present.

**Comment by groenator** [comment]
@hussfelt when you pack your things, did you used packer by any chance? I am using packer to build my AMI.

**Comment by groenator** [comment]
> @groenator if there is any help in this I had a problem where when I packaged my things I never cleared out the old packages. That meant that both idna 2.8 and idna 2.7 (which I replaced it with) was present.

@hussfelt when you pack your things, did you used packer by any chance? I am using packer to build my AMI.

**Comment by groenator** [comment]
I just found out that the issue it is because of the requests package. They
are going to release a new patch sometime next week.

Thanks!

**Comment by fwaechter** [comment]
For the Google/DuckDuckGo/Whatever users:

The Problem is that Debians version of the python2/3-cryptography package doesn't solve the requirements of the certbot version when certbot is installed with pip2/3. So the solution is quite simple: Remove Debians version of the python2/3-cryptography package and "reinstall" it with pip2/3. I guess it would also work the other way but I haven't tested it.

```
apt remove python3-cryptography
pip2/3 install cryptography
```
